src/scripts/sync_subclassof.py

```diff
@@ -0,0 +1,104 @@
+"""Create a robot template for purpose of syncing subclassof relations.
+
+Resources
+- GitHub issue: https://github.com/monarch-initiative/mondo-ingest/issues/92
+- GitHub PR: https://github.com/monarch-initiative/mondo-ingest/pull/363
+"""
+import os
+from argparse import ArgumentParser
+from pathlib import Path
+from typing import Any, Dict, List, Union
+
+import pandas as pd
+from oaklib import get_adapter
+from oaklib.interfaces.basic_ontology_interface import RELATIONSHIP
+from oaklib.types import CURIE, URI
+
+from src.scripts.utils import DEFAULT_PREFIXES_CSV, get_monarch_curies_converter, remove_angle_brackets
+
+HERE = Path(os.path.abspath(os.path.dirname(__file__)))
+SRC_DIR = HERE.parent
+ONTOLOGY_DIR = SRC_DIR / 'ontology'
+DEFAULT_OUTPATH = ONTOLOGY_DIR / 'reports' / 'sync-subclassof.robot.template.tsv'
+DEFAULT_ONTOLOGY_DB_PATH = ONTOLOGY_DIR / 'tmp' / 'merged.db'
+ROBOT_ROW = [{
+    'term_id': 'ID',
+    'parent_id': 'SC % SPLIT=|',
+    # todo: PR comment to remind about better property for:
+    'xref_ids': 'A oboInOwl:hasDbXref SPLIT=|',  # todo: Nico may provide different value
+}]
+
+# todo: oak issue: too slow iterating. 36 seconds for below snippet
+# for mondo_id in mondo_ids[0:10]:
+#     rels: List[CURIE] = [x for x in oi.relationships(subjects=[mondo_id])]
+
+# todo: add caching?
+def sync_subclassof(
+    outpath: str = DEFAULT_OUTPATH, prefixes_csv_path: str = DEFAULT_PREFIXES_CSV,
+    ontology_db_path: str = DEFAULT_ONTOLOGY_DB_PATH
+):
+    """Run"""
+    # Vars
+    conv = get_monarch_curies_converter(prefixes_csv_path)
+    compress = lambda x: conv.compress(x) if conv.compress(x) else x
+    oi = get_adapter(ontology_db_path)
+
+    # Get all terms
+    ids_all: List[Union[CURIE, URI]] = [x for x in oi.entities(filter_obsoletes=False)]
+    mondo_ids: List[CURIE] = [x for x in ids_all if x.startswith('MONDO:')]
+
+    # todo: pr comment?: skip if parent_id not in mondo? is this only the case for: [('MONDO:0000001', 'rdfs:subClassOf', 'BFO:0000016')]
+    #  TODO: analyze first
+    #   example: mondo:0000001 rdfs:subClassOf BFO:0000016
+    # Relationships
+    # todo: PR comment: how did i implement this and why. memory vs speed. timeout on all rels. 36 seconds on iter 1x/time for just 10 records
+    # todo: oak issue: slow even batch rels (all preds). 5 seconds for 10. 25 for 100. when passed 5 preds, 1 second for 100. 36 sec for 25k~
+    rels_raw: List[RELATIONSHIP] = [x for x in oi.relationships(
```

Comment:
## `relationships()` performance (ok)
@cmungall Just wanted to ask you / see if there are performance optimizations that can be done.
This way is fine:  `[x for x in oi.relationships(subjects=ALL_MONDO_IDS)]` 36 seconds for ~25k IDs.
This way took same number of seconds (36sec) for 10 IDs:
```py
for mondo_id in mondo_ids[0:10]:
    rels: List[CURIE] = [x for x in oi.relationships(subjects=[mondo_id])]
```
It would be convenient sometimes to do the call per ID, but it's quite expensive. It's really not a big deal though. I just wanted to bring to your attention. If you think it should actually be performing faster, I can open up an issue. Otherwise, can mark this comment 'resolved'.

Also, for some reason `[x for x in oi.relationships()]` (not passing any `subjects`) timed out. Maybe I'll try again.

Reply:
## `relationships()` strangeness?
@cmungall @hrshdhgd 

### 1. Default values
`[x for x in oi.relationships(subjects=['MONDO:0000087'])]`
Yields: `[('MONDO:0000087', 'RO:0002573', 'MONDO:0021136'), ('MONDO:0000087', 'rdfs:subClassOf', 'MONDO:0002320')]`
Question: It looks only a subset of predicates will return by default (e.g. `RO` preds and `rdfs:subClassOf`)? Is this intended behavior? If so, can we update the docstring for `.relationships()`. Currently in `basic_ontology_interface.py` it just reads: `:param predicates: constrain search to these predicates`

### 2. Missing `oboInOwl:hasDbXref`
`[x for x in oi.relationships(subjects=['MONDO:0000087'], predicates=['rdfs:subClassOf', 'oboInOwl:hasDbXref', 'skos:exactMatch', 'skos:broadMatch', 'skos:narrowMatch'])]`
Yields: `[('MONDO:0000087', 'rdfs:subClassOf', 'MONDO:0002320'), ('MONDO:0000087', 'skos:exactMatch', '<http://identifiers.org/mesh/D065706>'), ('MONDO:0000087', 'skos:exactMatch', '<http://identifiers.org/snomedct/4945003>'), ('MONDO:0000087', 'skos:exactMatch', '<http://linkedlifedata.com/resource/umls/id/C0266464>'), ('MONDO:0000087', 'skos:exactMatch', '<http://www.orpha.net/ORDO/Orphanet_35981>'), ('MONDO:0000087', 'skos:exactMatch', 'NCIT:C116936'), ('MONDO:0000087', 'skos:narrowMatch', '<https://icd.codes/icd10cm/Q04.3>')]`
Question: Is this a bug? I asked for `oboInOwl:hasDbXref` but didn't get it in the return. But in `merged.owl` for the `merged.db` I'm querying, there exists some of those (see snippet below). If you want, I can attach copies of the files, or re-create `merged.db` just to be sure it's up-to-date.

<details><summary>OWL for MONDO:0000087</summary>
<p>

```owl
    <owl:Class rdf:about="http://purl.obolibrary.org/obo/MONDO_0000087">
        <rdfs:subClassOf rdf:resource="http://purl.obolibrary.org/obo/MONDO_0002320"/>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://purl.obolibrary.org/obo/RO_0002573"/>
                <owl:someValuesFrom rdf:resource="http://purl.obolibrary.org/obo/MONDO_0021136"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <obo:IAO_0000115 rdf:datatype="http://www.w3.org/2001/XMLSchema#string">A developmental brain abnormality characterized by an excessive amount of small convolutions on the surface of the brain and cognitive dysfunction.</obo:IAO_0000115>
        <mondo:excluded_subClassOf rdf:resource="http://purl.obolibrary.org/obo/MONDO_0000508"/>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">GARD:0012271</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">ICD10CM:Q04.3</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">MESH:D065706</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">NCIT:C116936</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Orphanet:35981</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">SCTID:4945003</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">UMLS:C0266464</oboInOwl:hasDbXref>
        <oboInOwl:id rdf:datatype="http://www.w3.org/2001/XMLSchema#string">MONDO:0000087</oboInOwl:id>
        <oboInOwl:inSubset rdf:resource="http://purl.obolibrary.org/obo/mondo#disease_grouping"/>
        <oboInOwl:inSubset rdf:resource="http://purl.obolibrary.org/obo/mondo#ordo_group_of_disorders"/>
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">polymicrogyria</rdfs:label>
        <skos:exactMatch rdf:resource="http://identifiers.org/mesh/D065706"/>
        <skos:exactMatch rdf:resource="http://identifiers.org/snomedct/4945003"/>
        <skos:exactMatch rdf:resource="http://linkedlifedata.com/resource/umls/id/C0266464"/>
        <skos:exactMatch rdf:resource="http://purl.obolibrary.org/obo/NCIT_C116936"/>
        <skos:exactMatch rdf:resource="http://www.orpha.net/ORDO/Orphanet_35981"/>
        <skos:narrowMatch rdf:resource="https://icd.codes/icd10cm/Q04.3"/>
    </owl:Class>
```

</p>
</details>

Reply:
Will do a PR, I don't mind.
The 2nd column of the robot template is only `rdfs:subClassOf`.
The 3rd column of the template is [for xrefs](https://github.com/monarch-initiative/mondo-ingest/pull/363#discussion_r1328020237).
However, I'm also using `.relationships()` because it's the best way to [get at `obo:mondo#excluded_subClassOf`](https://github.com/monarch-initiative/mondo-ingest/pull/363#discussion_r1328031211).

Reply:
> However, I'm also using .relationships() because it's the best way to https://github.com/monarch-initiative/mondo-ingest/pull/363#discussion_r1328031211.

Ahhh didnt think of that! Good thinking!

Reply:
I think [chris responded to](https://github.com/monarch-initiative/mondo-ingest/pull/363#issuecomment-1730790535) "relationships() performance (ok)" with:
> Filter by press
> 
> Materialize views

But this seems very oracle-esque to me. Struggling to to read the tea leaves on this one.

Reply:
Note to self: Make a gh issue in oak for this and resolve

Reply:
Created an issue: https://github.com/INCATools/ontology-access-kit/issues/675